### PR TITLE
Thread registry

### PR DESCRIPTION
**Pull request by projedi**

### CODE REVIEW

File: runtime/src/mm/cpp/Memory.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+struct MemoryState {
+    kotlin::mm::ThreadData data;
+
+    ALWAYS_INLINE static MemoryState* from(kotlin::mm::ThreadData* data) { return wrapper_cast(MemoryState, data, data); }
+};
+
+MemoryState* InitMemory() {
+    auto* data = kotlin::mm::ThreadRegistry::instance().RegisterCurrentThread();
+    return MemoryState::from(data);
```
Comment (SvyatoslavScherbina) [inline runtime/src/mm/cpp/Memory.cpp:22]: Somewhat fragile. It makes some sense to add static assertions checking there are no more fields in MemoryState.

File: runtime/src/main/cpp/SingleLockList.hpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        LockGuard<Mutex> guard(mutex_);
+        if (root_) {
+            root_->previous = node.get();
+        }
+        node->next = std::move(root_);
+        root_ = std::move(node);
+        return result;
+    }
+
+    // You can only `Erase` `Value`s that were returned by `Emplace`. Trying
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/SingleLockList.hpp:70]: Would this be more type safe if we returned pointer to `Node` from `Emplace` instead?

File: runtime/src/mm/cpp/Memory.cpp (hunk 1 above)
Comment (projedi) [reply runtime/src/mm/cpp/Memory.cpp]: Left a single member inside `MemoryState` and left a comment explaining that it's a wrapper and asking not to add anything else to it. Should be sufficient.

File: runtime/src/main/cpp/SingleLockList.hpp (hunk 2 above)
Comment (projedi) [reply runtime/src/main/cpp/SingleLockList.hpp]: Done, less magic is involved now.

**Review by SvyatoslavScherbina — APPROVED** [review]

**Review by ilmat192 — APPROVED** [review]

**Pull request merged**